File: 290.word-pattern.java
public class Solution {
    public boolean wordPattern(String pattern, String str) {
        String[] strs = str.split(" ");
        if (pattern.length() != strs.length)
            return false;
        HashMap<Character, String> hm1 = new HashMap<Character, String>();
        HashMap<String, Character> hm2 = new HashMap<String, Character>();
        for (int i = 0; i < pattern.length(); ++i) {
            if (hm1.containsKey(pattern.charAt(i))) {
                if (!hm1.get(pattern.charAt(i)).equals(strs[i]))
                    return false;
            } else {
                if (hm2.containsKey(strs[i]))
                    return false;
                else {
                    hm1.put(pattern.charAt(i), strs[i]);
                    hm2.put(strs[i], pattern.charAt(i));
                }
            }
        }
        return true;
    }
}

File: 290.word-pattern.js
/*
 * @lc app=leetcode id=290 lang=javascript
 *
 * [290] Word Pattern
 */

// @lc code=start
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
/* Solution 1: Two Hash Map */
var wordPattern = function (pattern, str) {
  var pArr = pattern.split("");
  var sArr = str.split(" ");
  if (pArr.length !== sArr.length) return false;
  var mapP2S = {};
  var mapS2P = {};
  for (var i = 0; i < pArr.length; ++i) {
    if (mapP2S[pArr[i]]) {
      if (sArr[i] !== mapP2S[pArr[i]]) return false;
    } else {
      mapP2S[pArr[i]] = sArr[i];
    }
    if (mapS2P[sArr[i]]) {
      if (pArr[i] !== mapS2P[sArr[i]]) return false;
    } else {
      mapS2P[sArr[i]] = pArr[i];
    }
  }
  return true;
};
// @lc code=end
